### PR TITLE
Preserve state when using placeholders

### PR DESCRIPTION
**Pull request by andrewiggins**
Add a test and fix a bug where `null`, `false`, and `true` were removed from the render result before going into `diffChildren`. Calling `toChildArray` without a callback removes the values mentioned above, which we were doing on the result of render before going into `diffChildren`.

### CODE REVIEW

File: src/diff/index.js [hunk 1]
```diff
@@ -165,9 +165,7 @@ export function diff(
 			tmp = c.render(c.props, c.state, c.context);
 			let isTopLevelFragment =
 				tmp != null && tmp.type == Fragment && tmp.key == null;
-			newVNode._children = toChildArray(
-				isTopLevelFragment ? tmp.props.children : tmp
-			);
+			newVNode._children = isTopLevelFragment ? tmp.props.children : tmp;
```
Comment (andrewiggins) [inline src/diff/index.js]: Should this property be renamed to `_renderResult` since it isn't guaranteed to be an array until after `diffChildren` is complete? I vaguely remember this confusing people before. I was hesitant to do this immediately cuz it adds yet another property to VNode and this one is particularly short lived (it's just a way to pass the result into `diffChildren`)

File: test/browser/lifecycles/shouldComponentUpdate.test.js [hunk 2]
```diff
@@ -707,4 +707,24 @@ describe('Lifecycle methods', () => {
 			expect(scratch.textContent).to.equal('foo');
 		});
 	});
+
+	it('should correctly render when sCU component has null children', () => {
```
Comment (andrewiggins) [inline test/browser/lifecycles/shouldComponentUpdate.test.js]: This test was in this branch from a while ago. Not sure what it was trying to do but figured it can't hurt 🤷‍♂
